Better trash and fetch logic

nimmolo opened the pull request:
Silvia - I re-added all your commits one by one, and the plugin is working. Note:

1. `external-videos.php` Minor tweak - simplified the function `get_options` so the defaults are not defined twice.
2. `ev-admin.php` Changed functions relating to updating and deleting videos from authors to fix the persistent trash error with simpler logic. This one change affects four other functions, but i think you will agree it is an improvement. Here is an explanation:

`get_authors()` - removed this redundant function that was confusing the update and delete-videos loops. We don't need a list of all authors for any of the functions -we're always looping authors by host.

The functions
- `update_videos_handler`
- `post_new_videos`
- `trash_deleted_videos`
- `fetch_new_videos`

...now pass these simplified variables: 
- `$update_hosts` (**single host** if we're updating a single author, **all hosts** if updating all)
- `$update_author` (**single author** if we're updating a single author, **null** if updating all)

Most importantly, `trash_deleted_videos` still had a tortuous logic, now simplified:

When fetching the updated array of videos from a single author's hosting service, the function was still looping through an array of **all existing external-videos posts** and comparing with the author's updated array of $new_videos. Anything in $all_existing_videos that was not in the single author's $new_videos was assumed to be "deleted on the host", so it was trashed. 

You wrote a fix to skip to next video if the host was not an $update_hosts, 
```
// next video if we're looking at a video not on a updated site
if ( !in_array( $host, $update_hosts ) ) continue;
```
but if there were **two authors on the same host**, this would still delete the other author's videos during a single author update - they would not be in the single author's $new_videos array, even though the $host was the same. 

It seemed simpler for `trash_deleted_videos` to have two versions of `WP_Query` to compare existing videos on the WP instance with updated videos on the host. 
- If we're updating a single author, it uses a narrower `WP_Query` with a meta query, and only loops the **existing videos of the relevant author** to check against $new_videos. 
- If updating all, it loops all external-videos as before.

nimmolo committed:
Merge remote-tracking branch 'silviapfeiffer/master'

nimmolo committed:
backtrace

Added all Silvia’s commits.
Overwrote the following functions in ev-admin.php:

get_authors() - removed redundant function that confused the $host loop

the following functions now pass simplified variables: $update_hosts
and $update_author (specified if updating single author, null if
updating all)

update_videos_handler
post_new_videos
trash_deleted_videos
fetch_new_videos

trash_deleted_videos now has refined conditional wp_query, so if
updating single author, it is only checking new_videos against the
videos of the relevant author

silviapfeiffer commented on core/ev-admin.php:
```diff
@@ -253,20 +223,20 @@ function update_videos_handler() {
   *  Saves any new videos from host channels to the database.
   *  Returns messages about number of video posts added.
-  *  Works for single-author and update-all
+  *  Works for single-author and update-all via $single param
```
$single is still not used, so this addition is not necessary

nimmolo replied:
fixed

silviapfeiffer commented on core/ev-admin.php:
```diff
@@ -392,47 +383,49 @@ function trash_deleted_videos( $update_hosts, $new_video_ids ) {
 
-    return '';
+    // return the messages
+    return $trash_messages;
```
If you get here, $trash_messages is not set, so this will fail. You need to return the $trash_messages from within the isset() if-statement and return empty here.

nimmolo replied:
right!

silviapfeiffer commented on external-videos.php:
```diff
@@ -223,12 +222,9 @@ function admin_scripts( $hook ) {
     wp_register_style( 'ev-admin', plugin_dir_url( __FILE__ ) . '/css/ev-admin.css', array(), null, 'all' );
     wp_enqueue_style( 'ev-admin' );
-    wp_register_script( 'ev-admin', plugin_dir_url( __FILE__ ) . '/js/ev-admin.js', array( 'jquery'/*, 'jquery-ui-tabs', 'jquery-ui-accordion' */), false, true );
+    wp_register_script( 'ev-admin', plugin_dir_url( __FILE__ ) . '/js/ev-admin.js', array( 'jquery' ), false, true );
```
Hmm, interesting - does that fix the settings page?

nimmolo replied:
I don't know - it worked with the commented stuff too!  Anyway it works now

silviapfeiffer commented on external-videos.php:
```diff
@@ -259,26 +255,25 @@ function admin_scripts( $hook ) {
+
     // Set defaults for the basic options
-    if( !$raw_options ) {
-      $options = array( 'version' => 1, 'rss' => false, 'delete' => true, 'hosts' => array(), 'attrib' => false, 'loop' => false, 'slug' => 'external-videos');
```
You should not remove this - this is initialization for the first time somebody installs the plugin!

nimmolo replied:
This I tested - it is OK to remove actually. 

At this point, for first time initialization, $options is now set a blank array (line 262), and the isset() statements below set our defaults one key at a time. The reason one-at-a-time is preferable is for upgrades - people with earlier versions will have some of the keys set, but not the new ones. So upgraders don't get the defaults unless it checks for them one by one.

But definitely, first time users get all defaults set by the isset() statements. I just tested on a fresh install.

silviapfeiffer reviewed (changes requested):
A few things to check.

nimmolo commented:
just woke up and realized - my resubmitted PR is probably missing your fixes on readme and  one of the hosts. will fix tomorrow and resubmit (addressing these other things) - 3:48am going back to sleep!

nimmolo commented:
OK now i'm awake! And i realize this resubmitted PR did include your fixes on readme, forms, vimeo and wistia. Just double-checked that now. Committing these changes...

nimmolo committed:
Requested changes

Also: dotsub embed code fix. This should be a separate PR, sorry

nimmolo commented:
Silvia - I've asked Dominik Schilling of the WP security team to review the plugin for XSS holes. I just sent him an email - don't know what he'll say, but if he says yes, is it OK with you if I send him a Github review request?

silviapfeiffer commented:
Sure! I don't think he'll do it - if every plugin asked him to check, it would take 100 of him to do his job. :-)

silviapfeiffer commented:
Also, sorry I haven't landed this yet - will need to wait until the weekend.

nimmolo commented:
Client asked for one more feature - a post-status select, to choose if external-videos are automatically Published or set to Draft, Private, Pending.  I will send this via a separate PR.

nimmolo commented:
Nevermind - delirious ! forgot post_status setting exists already

nimmolo committed:
Readme updates

Describe new settings for loop and slug in FAQ
Describe setting for post_status

nimmolo commented:
Silvia should we merge this PR so my subsequent PRs will not appear to have more diffs than they do?

I have one more change - host classes should use their own embed_code functions when building the post_meta for each video, so it's not defined inconsistently.

nimmolo committed:
Thorough testing changes

1. sanitize and escape all values coming from video host query
2. standardize `$video` array keys to be the same as the eventual
post_meta keys (i.e. ~~['videourl']~~ ['video_url'], ~~['thumbnail']~~
['thumbnail_url'] ), just for consistency and sanity
3. break out the function that standardizes each host's video data into
a separate function for easier reading, `compose_video( $vid )`
4. changed the name of the `embed_code()` function to `embed_url()` and
reuse `embed_url()` function within the `compose_video` function,
rather than duplicating code
6. all hosts: fix embed codes, duration, category and tags (should
always be array)
7. all hosts: make admin host instructions translatable
8. huge advance: pull 'unobtainable' YouTube duration and tags from
unprecedented fourth API query
9. Dotsub and Wistia and Dailymotion: add FitVids notice
10. Added Youtube duration conversion helper function

nimmolo commented:
I imagine you're thoroughly skeptical of my Quality Control at this point ;) but i must say i've really scratched every corner of the plugin anew this time, and fixed many things that weren't working, most of which I never noticed before. I put in a good 10 hour day on it today.

I paid very close attention to all the processes by which an external-videos **post is created**, and how the **fields are filled out** and whether the end values **correspond to host data**. I had many errors in my API correspondence with different hosts, from incorrectly parsed video duration time strings, to messed up embed codes. WP for instance is quite lenient with youtube embeds, but now the use of each host's "embed code" is strictly followed in the post creation now, as well. At youtube, i also located the final hurdle that gives access to the mysterious "video duration" and "tags" fields - you have to do yet another API call for every video, after you have the ID, to get this info, but now we have it. And it does make for more useful posts. 

I made a **major effort with sanitization** too. I sanitized anything that gets put into the database during save_post, and redundantly sanitized everything coming from the host APIs as well. With the amount of XSS happening out there it seems wise to be as safe as possible - for example, a spoof of an API endpoint seems conceivable.

There's a bit more code restructuring too, some for my sanity, and some to prepare for the future media explorer update.  

I think we're in better shape now - I didn't get any errors all the way through the processes of adding authors, updating videos, deleting videos, and posting videos, for every host. 

I made the host field explanations translatable, too. Other things listed on the last commit!

silviapfeiffer commented:
Wow, thanks for the write-up - stuff like that should be in commit messages! Anyway, I just haven't had time to check it! I'm getting to it this weekend, promised!

silviapfeiffer commented on readme.md:
```diff
@@ -74,6 +74,18 @@ And you can specify `[external-videos link="page"]` if you want to get the links
 Just add the following to your Website URL: ?feed=rss2&post_type=external-videos .
 You can add a link like this to your theme layout.
 
+### How can I get external videos to appear in the Home page ("page for posts")?
```
Awesome!
